stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -192,58 +186,74 @@ function Base.show(io::IO, ::MIME"text/plain", S::AbstractSparseMatrixCSC)
     m, n = size(S)
     print(io, m, "×", n, " ", typeof(S), " with ", xnnz, " stored ",
               xnnz == 1 ? "entry" : "entries")
-    if xnnz != 0
+    if !(m == 0 || n == 0)
         print(io, ":")
         show(IOContext(io, :typeinfo => eltype(S)), S)
     end
 end
 
 Base.show(io::IO, S::AbstractSparseMatrixCSC) = Base.show(convert(IOContext, io), S::AbstractSparseMatrixCSC)
-function Base.show(io::IOContext, S::AbstractSparseMatrixCSC)
-    nnz(S) == 0 && return show(io, MIME("text/plain"), S)
-
-    ioc = IOContext(io, :compact => true)
-    function _format_line(r, col, padr, padc)
-        print(ioc, "\n  [", rpad(rowvals(S)[r], padr), ", ", lpad(col, padc), "]  =  ")
-        if isassigned(nonzeros(S), Int(r))
-            show(ioc, nonzeros(S)[r])
-        else
-            print(ioc, Base.undef_ref_str)
-        end
-    end
 
-    function _get_cols(from, to)
-        idx = eltype(getcolptr(S))[]
-        c = searchsortedlast(getcolptr(S), from)
-        for i = from:to
-            while i == getcolptr(S)[c+1]
-                c +=1
-            end
-            push!(idx, c)
-        end
-        idx
-    end
-
-    rows = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
-    if !get(io, :limit, false) || rows >= nnz(S) # Will the whole matrix fit when printed?
-        cols = _get_cols(1, nnz(S))
-        padr, padc = ndigits.((maximum(rowvals(S)[1:nnz(S)]), cols[end]))
-        _format_line.(1:nnz(S), cols, padr, padc)
-    else
-        if rows <= 2
-            print(io, "\n  \u22ee")
-            return
-        end
-        s1, e1 = 1, div(rows - 1, 2) # -1 accounts for \vdots
-        s2, e2 = nnz(S) - (rows - 1 - e1) + 1, nnz(S)
-        cols1, cols2 = _get_cols(s1, e1), _get_cols(s2, e2)
-        padr = ndigits(max(maximum(rowvals(S)[s1:e1]), maximum(rowvals(S)[s2:e2])))
-        padc = ndigits(cols2[end])
-        _format_line.(s1:e1, cols1, padr, padc)
-        print(io, "\n  \u22ee")
-        _format_line.(s2:e2, cols2, padr, padc)
-    end
-    return
+const brailleBlocks = UInt16['⠁', '⠂', '⠄', '⡀', '⠈', '⠐', '⠠', '⢀']
+function Base.show(io::IOContext, S::AbstractSparseMatrixCSC)
+    m, n = size(S)
+    (m == 0 || n == 0) && return show(io, MIME("text/plain"), S)
+
+    # The maximal number of characters we allow to display the matrix
+    maxHeight = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
+    maxWidth = displaysize(io)[2] ÷ 2
```

Comment:
`displaysize(io::IOContext)` returns `Any` due to the branch in https://github.com/JuliaLang/julia/blob/05a80e1e82ac976bd3323f10e1139d86b3954cbe/base/show.jl#L336
As a result, substantial portions of the code below get "poisoned" and is inferred as `Any`, including calls to `brailleBlocks[p]` since `p` is also inferred as `Any`. This adds a lot of allocations to this method.
I guess this problem existed before this PR, but I thought I'd point it out anyway.

Reply:
Is there a way to solve this issue, maybe by doing
```julia
maxHeight = convert(Int, displaysize(io)[1] - 4)
maxWidth = convert(Int, displaysize(io)[2] ÷ 2)
```
?

Reply:
You can add a type declaration
```
local maxHeight::Int 
local maxWidth::Int 
maxHeight, maxWidth = displaysize(io)
```